### PR TITLE
Esb 226

### PR DESCRIPTION
**Pull request by solglass**
Lesson controller validation

### CODE REVIEW

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 1]
```diff
@@ -17,26 +21,39 @@ namespace EducationSystem.Controllers
     public class LessonController : ControllerBase
     {
         private ILessonService _lessonService;
+        private IGroupService _groupService;
+        private ICourseService _courseService;
         private IMapper _mapper;
+        private int _userId;
+        private List<int> _userGroups;
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:28]: do we really need these two fields?

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 2]
```diff
@@ -46,13 +63,25 @@ public ActionResult<LessonOutputModel> AddNewLesson([FromBody] LessonInputModel
...
-        public ActionResult<List<LessonOutputModel>> GetLessons(int groupId)
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        // https://localhost:50221/api/lesson/by-group/45
+        [HttpGet("by-group/{groupId}")]
+        [Authorize(Roles = "Администратор, Преподаватель")]
+        public ActionResult<List<LessonOutputModel>> GetLessonsByGroupId(int groupId)
         {
+
+            _userGroups = ControllerBaseExtension.SupplyUserGroupsList((ControllerBase) this,_groupService);
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:74]: please simplify it to "this.SupplyUserGroupsList(_groupService)" as it done below

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 3]
```diff
@@ -46,13 +63,25 @@ public ActionResult<LessonOutputModel> AddNewLesson([FromBody] LessonInputModel
...
+            var group = _groupService.GetGroupById(groupId);
+            if (group == null)
+                return NotFound($"Group {groupId} was not found");
+
+            if (!(_userGroups.Contains(groupId)))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {groupId}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:80]: please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 4]
```diff
@@ -46,13 +63,25 @@ public ActionResult<LessonOutputModel> AddNewLesson([FromBody] LessonInputModel
...
+        [Authorize(Roles = "Администратор, Преподаватель")]
+        public ActionResult<List<LessonOutputModel>> GetLessonsByGroupId(int groupId)
         {
+
+            _userGroups = ControllerBaseExtension.SupplyUserGroupsList((ControllerBase) this,_groupService);
+            var group = _groupService.GetGroupById(groupId);
+            if (group == null)
+                return NotFound($"Group {groupId} was not found");
+
+            if (!(_userGroups.Contains(groupId)))
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:79]: so "Администратор" will be always receiving the 403rd, right?)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 5]
```diff
@@ -62,13 +91,25 @@ public ActionResult<List<LessonOutputModel>> GetLessons(int groupId)
...
         public ActionResult<LessonOutputModel> GetLessonById(int lessonId)
         {
+           _userGroups = this.SupplyUserGroupsList(_groupService);
+
             var lessonDto = _lessonService.GetLessonById(lessonId);
+
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:108]: please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 6]
```diff
@@ -62,13 +91,25 @@ public ActionResult<List<LessonOutputModel>> GetLessons(int groupId)
...
+        [Authorize(Roles = "Администратор, Преподаватель, Студент")]
         public ActionResult<LessonOutputModel> GetLessonById(int lessonId)
         {
+           _userGroups = this.SupplyUserGroupsList(_groupService);
+
             var lessonDto = _lessonService.GetLessonById(lessonId);
+
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:107]: wrong condition

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 7]
```diff
@@ -77,12 +118,20 @@ public ActionResult<LessonOutputModel> GetLessonById(int lessonId)
...
+        [Authorize(Roles = "Администратор, Преподаватель")]
         public ActionResult<LessonOutputModel> DeleteLesson(int lessonId)
         {
-           _lessonService.DeleteLesson(lessonId);
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:133]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 8]
```diff
@@ -77,12 +118,20 @@ public ActionResult<LessonOutputModel> GetLessonById(int lessonId)
         /// <param name="lessonId"> Id of the lesson</param>
         /// <returns>Output model of the soft-deleted Lesson</returns>
         [ProducesResponseType(typeof(LessonOutputModel), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(typeof(LessonOutputModel), StatusCodes.Status404NotFound)]
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:122]: please remove typeof

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 9]
```diff
@@ -93,12 +142,20 @@ public ActionResult<LessonOutputModel> DeleteLesson(int lessonId)
         /// <param name="lessonId"> Id of the lesson</param>
         /// <returns>Output model of the recovered Lesson</returns>
         [ProducesResponseType(typeof(LessonOutputModel), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(typeof(LessonOutputModel), StatusCodes.Status404NotFound)]
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:146]: please remove typeof

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 10]
```diff
@@ -93,12 +142,20 @@ public ActionResult<LessonOutputModel> DeleteLesson(int lessonId)
...
+        [Authorize(Roles = "Администратор, Преподаватель")]
         public ActionResult<LessonOutputModel> RecoverLesson(int lessonId)
         {
-             _lessonService.RecoverLesson(lessonId);
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:157]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 11]
```diff
@@ -107,17 +164,27 @@ public ActionResult<LessonOutputModel> RecoverLesson(int lessonId)
...
         {
-            var lessonDto = _mapper.Map<LessonDto>(inputModel);
+            if (!ModelState.IsValid)
+                throw new ValidationException(ModelState);
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:184]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 12]
```diff
@@ -107,17 +164,27 @@ public ActionResult<LessonOutputModel> RecoverLesson(int lessonId)
         /// Updates lesson's properties
         /// </summary>
         /// <param name="lessonId">Id of the lesson</param>
-         /// <param name="inputModel">Input model with all the properties for the new lesson</param>
+        /// <param name="inputModel">Input model with all the properties for the new lesson</param>
         /// <returns>Output model of the updated Lesson</returns>
         [ProducesResponseType(typeof(LessonOutputModel), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status409Conflict)]
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:171]: where is 404?

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 13]
```diff
@@ -129,11 +196,28 @@ public ActionResult<LessonOutputModel> UpdateLesson(int lessonId, [FromBody] Les
...
+                throw new ValidationException(ModelState);
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            var groupDto = _groupService.GetGroupById((int)inputModel.GroupId);
+            if (groupDto == null)
+                return NotFound($"Group {groupDto} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:218]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 14]
```diff
@@ -129,11 +196,28 @@ public ActionResult<LessonOutputModel> UpdateLesson(int lessonId, [FromBody] Les
...
+            if (lessonDto.Group.Id != (int)inputModel.GroupId)
+                return BadRequest($"Group with id {(int)inputModel.GroupId} does not belong to the lesson with id {lessonId}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:220]: group is a kind of "master" entity, so the lesson doesn't belong to group

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 15]
```diff
@@ -146,12 +230,26 @@ public ActionResult<List<FeedbackOutputModel>> GetFeedbacks(int lessonId, [FromB
...
-        [Authorize(Roles = "Админ, Менеджер, Методист, Преподаватель")]
+        [Authorize(Roles = "Администратор, Менеджер, Методист, Преподаватель")]
         public ActionResult<FeedbackOutputModel> GetFeedbackById(int lessonId, int feedbackId)
         {
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:246]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 16]
```diff
@@ -164,11 +262,24 @@ public ActionResult<FeedbackOutputModel> GetFeedbackById(int lessonId, int feedb
...
         {
+            if (!ModelState.IsValid)
+                throw new ValidationException(ModelState);
+
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:281]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 17]
```diff
@@ -182,12 +293,37 @@ public ActionResult<FeedbackOutputModel> AddNewFeedback(int lessonId, FeedbackIn
...
-            var feedbackDto = _mapper.Map<FeedbackDto>(feedbackInputModel);
+            if (!ModelState.IsValid)
+                throw new ValidationException(ModelState);
+
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:313]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 18]
```diff
@@ -182,12 +293,37 @@ public ActionResult<FeedbackOutputModel> AddNewFeedback(int lessonId, FeedbackIn
...
+            var feedbackDto = _lessonService.GetFeedbackById(feedbackId);
+            if (feedbackDto == null)
+                return NotFound($"Feedback {feedbackId} was not found");
+            if (feedbackDto.Lesson.Id != lessonId)
+                return BadRequest($"Feedback with id {feedbackId} does not belong to the lesson with id {lessonId}");
+            if (!(feedbackDto.User.Id == _userId) && !(User.IsInRole("Администратор")))
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:319]: !(feedbackDto.User.Id == _userId) => feedbackDto.User.Id != _userId

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 19]
```diff
@@ -182,12 +293,37 @@ public ActionResult<FeedbackOutputModel> AddNewFeedback(int lessonId, FeedbackIn
...
+                return StatusCode(StatusCodes.Status403Forbidden, $"User cannot update feedback {feedbackDto.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:320]: please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 20]
```diff
@@ -200,12 +336,30 @@ public ActionResult<FeedbackOutputModel> UpdateFeedback(int lessonId, int feedba
...
-        [Authorize(Roles = "Админ, Студент")]
+        [Authorize(Roles = "Администратор, Студент")]
         public ActionResult DeleteFeedback(int lessonId, int feedbackId)
         {
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:353]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 21]
```diff
@@ -200,12 +336,30 @@ public ActionResult<FeedbackOutputModel> UpdateFeedback(int lessonId, int feedba
...
+            var feedbackDto = _lessonService.GetFeedbackById(feedbackId);
+            if (feedbackDto == null)
+                return NotFound($"Feedback {feedbackId} was not found");
+            if (feedbackDto.Lesson.Id != lessonId)
+                return BadRequest($"Feedback with id {feedbackId} does not belong to the lesson with id {lessonId}");
+            if (!(feedbackDto.User.Id == _userId) && !(User.IsInRole("Администратор")))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User cannot delete feedback {feedbackDto.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:360]: 1. !(feedbackDto.User.Id == _userId) => feedbackDto.User.Id != _userId
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 22]
```diff
@@ -216,11 +370,20 @@ public ActionResult DeleteFeedback(int lessonId, int feedbackId)
...
-        [Authorize(Roles = "Админ, Преподаватель, Менеджер")]
+        [Authorize(Roles = "Администратор, Преподаватель, Менеджер")]
         public ActionResult<List<AttendanceOutputModel>> GetAttendancesByLessonId(int lessonId)
         {
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:385]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 23]
```diff
@@ -233,12 +396,25 @@ public ActionResult<List<AttendanceOutputModel>> GetAttendancesByLessonId(int le
...
-        [Authorize(Roles = "Админ, Преподаватель, Менеджер")]
+        [Authorize(Roles = "Администратор, Преподаватель, Менеджер")]
         public ActionResult<AttendanceOutputModel> GetAttendanceById(int lessonId, int attendanceId)
         {
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:412]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 24]
```diff
@@ -250,11 +426,24 @@ public ActionResult<AttendanceOutputModel> GetAttendanceById(int lessonId, int a
...
         {
+            if (!ModelState.IsValid)
+                throw new ValidationException(ModelState);
+
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:445]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 25]
```diff
@@ -270,11 +459,33 @@ public ActionResult<AttendanceOutputModel> GetAttendanceById(int lessonId, int a
...
-            var attendanceDto = _mapper.Map<AttendanceDto>(attendanceInputModel);
+            if (!ModelState.IsValid)
+                throw new ValidationException(ModelState);
+
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:479]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 26]
```diff
@@ -270,11 +459,33 @@ public ActionResult<AttendanceOutputModel> GetAttendanceById(int lessonId, int a
...
+            var attendanceDto = _lessonService.GetAttendanceById(attendanceId);
+            if (attendanceDto == null)
+                return NotFound($"Attendence {attendanceId} was not found");
+            if (attendanceDto.Lesson.Id != lessonId)
+                return BadRequest($"Attendence with id {attendanceId} does not belong to the lesson with id {lessonId}");
+            if (!(attendanceDto.User.Id == _userId) && !(User.IsInRole("Администратор")))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User cannot update attendence {attendanceDto.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:486]: 1. !(attendanceDto.User.Id == _userId) => attendanceDto.User.Id != _userId
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 27]
```diff
@@ -286,12 +497,30 @@ public ActionResult<AttendanceOutputModel> UpdateAttendance(int lessonId, int at
...
-        [Authorize(Roles = "Админ, Преподаватель")]
+        [Authorize(Roles = "Администратор, Преподаватель")]
         public ActionResult DeleteAttendance(int lessonId, int attendanceId)
         {
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:514]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 28]
```diff
@@ -286,12 +497,30 @@ public ActionResult<AttendanceOutputModel> UpdateAttendance(int lessonId, int at
...
+            var attendanceDto = _lessonService.GetAttendanceById(attendanceId);
+            if (attendanceDto == null)
+                return NotFound($"Attendence {attendanceId} was not found");
+            if (attendanceDto.Lesson.Id != lessonId)
+                return BadRequest($"Attendence with id {attendanceId} does not belong to the lesson with id {lessonId}");
+
+            if (!(attendanceDto.User.Id == _userId) && !(User.IsInRole("Администратор")))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User cannot delete  attendence {attendanceDto.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:522]: 1. !(attendanceDto.User.Id == _userId) => attendanceDto.User.Id != _userId
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 29]
```diff
@@ -304,10 +533,16 @@ public ActionResult DeleteAttendance(int lessonId, int attendanceId)
...
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
         [HttpGet("by-theme/{themeId}")]
-        [Authorize(Roles = "Админ, Преподаватель, Студент, Тьютор")]
+        [Authorize(Roles = "Администратор, Преподаватель, Тьютор")]
         public ActionResult<List<LessonOutputModel>> GetLessonsByThemeId(int themeId)
         {
+            var themeDto = _courseService.GetThemeById(themeId);
+            if (themeDto == null)
+                return NotFound($"Theme {themeId} was not found");
+
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:545]: dyren' in bezopasnost'
each teacher and each tutor gets the whole list of lessons (svoi, ne svoi - poher)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 30]
```diff
@@ -318,12 +553,25 @@ public ActionResult<List<LessonOutputModel>> GetLessonsByThemeId(int themeId)
...
-        [Authorize(Roles = "Админ, Преподаватель")]
+        [Authorize(Roles = "Администратор, Преподаватель")]
         public ActionResult AddNewLessonTheme(int lessonId, int themeId)
         {
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:570]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 31]
```diff
@@ -334,14 +582,31 @@ public ActionResult AddNewLessonTheme(int lessonId, int themeId)
...
+        [Authorize(Roles = "Администратор, Преподаватель")]
         public ActionResult DeleteLessonTheme(int lessonId, int themeId)
         {
-             _lessonService.DeleteLessonTheme(lessonId, themeId);
+            _userGroups = this.SupplyUserGroupsList(_groupService);
+            var lessonDto = _lessonService.GetLessonById(lessonId);
+            if (lessonDto == null)
+                return NotFound($"Lesson {lessonId} was not found");
+            if (!(User.IsInRole("Администратор") && !(_userGroups.Contains(lessonDto.Group.Id))))
+                return StatusCode(StatusCodes.Status403Forbidden, $"User is not in group {lessonDto.Group.Id}");
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:599]: 1. wrong condition
2. please change to return Forbid(...)

File: EducationSystem/EducationSystem/Controllers/LessonController.cs [hunk 32]
```diff
@@ -334,14 +582,31 @@ public ActionResult AddNewLessonTheme(int lessonId, int themeId)
...
+            var themeDto = _courseService.GetThemeById(themeId);
+            if (themeDto == null)
+                return NotFound($"Theme {themeId} was not found");
+            var themeFound = lessonDto.Themes.First(theme => theme.Id == themeDto.Id);
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Controllers/LessonController.cs:603]: possible NullReferenceException

File: EducationSystem/EducationSystem/Properties/launchSettings.json [hunk 33]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (Negodyaj) [inline EducationSystem/EducationSystem/Properties/launchSettings.json:1]: must be added to gitignore

**Comment by solglass** [comment]
Fixed after review

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 29 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs:549]: Re-written as a quickfix on a controller level due to the fact that Forbid() for every wrong lesson in this case is somewhat excessive in my opinion.        
var lessonDtos = _lessonService.GetLessonsByThemeId(themeId);
            lessonDtos.ForEach(lesson =>
            {
                if (!(User.IsInRole("Администратор")) && !(_userGroups.Contains(lesson.Group.Id)))
                    lessonDtos.Remove(lesson);
            });

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 32 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs]: changed to FirstOrDefault()

File: EducationSystem/EducationSystem/Properties/launchSettings.json (hunk 33 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Properties/launchSettings.json:1]: added adter dev merge into the branch

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 4 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs]: fixed             if (!(User.IsInRole("Администратор")) && !(_userGroups.Contains(groupId)))
                return Forbid($"User is not in group {groupId}");

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 5 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs]: done in this case and other

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 30 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs]: 1) Bracket misplaced 
if (!(User.IsInRole("Администратор")) && !(_userGroups.Contains(lessonDto.Group.Id))) 
should do in this case and other
2) Done

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 8 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs]: fixed

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 12 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs:170]: added

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 14 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs]: wording if the error message changed to ($"Lesson with id {lessonId} does not belong to the group with id {(int)inputModel.GroupId}")

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 18 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs]: changed

File: EducationSystem/EducationSystem/Controllers/LessonController.cs (hunk 21 above)
Comment (solglass) [reply EducationSystem/EducationSystem/Controllers/LessonController.cs]: changed

**Pull request merged**